source/h2_stream.c

```diff
@@ -871,6 +885,28 @@ struct aws_h2err aws_h2_stream_on_decoder_data_begin(
         return s_send_rst_and_close_stream(stream, aws_h2err_from_h2_code(AWS_HTTP2_ERR_PROTOCOL_ERROR));
     }
 
+    if (stream->thread_data.content_length != -1) {
+        stream->thread_data.received_data_length += payload_len; /* Should never exceed int64_t in real life */
+        if (end_stream && stream->thread_data.received_data_length != (uint64_t)stream->thread_data.content_length) {
```

Comment:
just use uint64's all around and a separate bool and the u64_checked() functions so I can stop worrying about overflow

Reply:
ALSO
check that we haven't exceeded it, even if !end_stream
AND
then if it is end_stream check that they're perfectly matched